plutus-conformance/add-test-output/Spec.hs

```diff
@@ -0,0 +1,116 @@
+runnerReader "eval" = Right Eval
+runnerReader "typecheck" = Right Typecheck
+runnerReader inp =
+  Left ("Unsupported test " <> show inp <>
+        ". Please choose either eval (for evaluation tests) or typecheck (for typechecking tests).")
+
+args :: ParserInfo Args
+args = info ((MkArgs <$> ext <*> dir <*> runner) <**> helper)
+  (fullDesc <> progDesc helpText)
+
+helpText :: String
+helpText = unlines
+  ["This program adds test outputs to specified inputs."
+  , "To run the program, input the following 3 arguments: "
+  , "(1) file extension to be searched "
+  , "(2) directory to be searched "
+  , "(3) the action to run the input files through; eval (for evaluation tests) or typecheck (for typechecking tests). "
+  , "E.g. run "
+  , "`cabal run add-test-output .uplc plutus-conformance/uplc/ eval` "
+  , "to have the executable search for files with extension `.uplc` in the /uplc directory that are missing output files. "
+  , " It will evaluate and create output files for them."
+  ]
+
+
+main :: IO ()
+main = do
+    MkArgs extension directory run <- customExecParser (prefs showHelpOnEmpty) args
+    allInputFiles <- findByExtension [extension] directory
+    -- only choose the ones without an output file, so as to not edit the ones already with outputs
+    inputFiles <- filterM (fmap (fmap not) (\testIn -> doesFileExist (testIn <> ".expected"))) allInputFiles
+    case run of
+      Eval -> do
+        for_ inputFiles
+          (\inputFile -> do
+            inputStr <- readFile inputFile
+            let parsed = runQuoteT $ UPLC.parse UPLC.program inputFile $ T.pack inputStr
+                outFilePath = inputFile <> ".expected"
+            case parsed of
+              Left (ParseErrorB peb) -> do
+                -- warn the user that the file failed to parse
+                putStrLn $ inputFile <> " failed to parse. Error written to " <> outFilePath
+                writeFile outFilePath (errorBundlePretty peb)
+              Right pro -> do
+                res <- try (evaluate $ evalUplcProg (() <$ pro)):: IO (Either SomeException (EvaluationResult UplcProg))
+                case res of
+                  Right (EvaluationSuccess prog) -> do
+                    putStrLn $ inputFile <> " evaluated; result written to " <> outFilePath
+                    writeFile outFilePath (render $ pretty prog)
+                  Right EvaluationFailure      -> do
+                    -- warn the user that the file failed to evaluate
+                    putStrLn $ inputFile <> " failed to evaluate. Failure written to " <> outFilePath
+                    writeFile outFilePath (show (EvaluationFailure :: EvaluationResult UplcProg))
+                  Left _ -> do
+                    -- warn the user that exception is thrown
+                    putStrLn $ "Exception thrown during evaluation of " <> inputFile <>".Written to " <> outFilePath
+                    writeFile outFilePath "Evaluation failure"
```

Comment:
It's `"Evaluation failure"` here, but `show EvaluationFailure` above? Seems inconsistent. I like the former option more, by the way. Probably even `"evaluation failure"` without the capital `E`, but that's inhuman amounts of nitpicking.

So I'd define `shownEvaluationFailure :: String` (or `Text` or whatever) and use that consistently.

Reply:
I will make a show instance for `EvaluationFailure` as "evaluation failure" (as you suggested). It'll be used more consistently that way I think.

Reply:
nvm, I'll make the `shownEvaluationFailure` function as you suggested. I know we prefer to use the derived show.